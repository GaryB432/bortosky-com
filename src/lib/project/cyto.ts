import type {
  EdgeDefinition,
  ElementsDefinition,
  NodeDefinition,
} from 'cytoscape';
import type { GaryProject } from './project';

export function asdf() {}

export async function getElements(
  gprojs: GaryProject[],
): Promise<ElementsDefinition> {
  const mns = new Map<string, NodeDefinition>();
  const mes = new Map<string, EdgeDefinition>();

  for (const gp of gprojs) {
    const id = gp.root.name;
    mns.set(id, { data: { id }, classes: ['gbp'] });
    for (const sp of gp.projects) {
      const sid = `${id}#${sp.name}`;
      mns.set(sid, { data: { id: sid }, classes: ['psub'] });
      const eid = `${sid}-${id}`;

      mes.set(eid, {
        data: {
          id: eid,
          source: id,
          target: sid,
        },
        classes: 'psub',
      });
    }

    Object.entries(gp.root.dependencies ?? {})
      .map(([k, v]) => {
        const did = `${k}@${id}`;
        return { data: { id: did, v }, classes: ['dep', 'prod'] };
      })
      .forEach((d) => {
        mns.set(d.data.id, d);
      });

    Object.entries(gp.root.devDependencies ?? {})
      .map(([k, v]) => {
        const did = `${k}@${id}`;
        return { data: { id: did, v }, classes: ['dep', 'dev'] };
      })
      .forEach((d) => {
        mns.set(d.data.id, d);
      });
  }

  const nodes = Array.from(mns.values()).sort((a, b) =>
    a.data.id ? a.data.id.localeCompare(b.data.id ?? '') : 0,
  );

  const edges = Array.from(mes.values()).sort((a, b) => 0);

  return { nodes, edges };
}
