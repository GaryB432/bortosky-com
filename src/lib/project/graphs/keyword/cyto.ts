import type { PackageJson } from "$lib/project/project";
import type {
  CssStyleDeclaration,
  EdgeDataDefinition,
  EdgeDefinition,
  ElementDataDefinition,
  ElementsDefinition,
  NodeDefinition,
} from "cytoscape";

type ElementDataWithId = Required<Pick<ElementDataDefinition, "id">>;

type PackageJsonDataDefinition = ElementDataWithId & PackageJson;

type KeywordDataDefinition = ElementDataWithId;

type PkEdgeDefinition = ElementDataWithId & EdgeDataDefinition;

export type GElementDataDefinitionx = ElementDataDefinition & {
  description?: string;
  id: string;
  label?: string;
};

// export async function getDependencyElements(
//   gprojs: GaryProject[],
//   filter?: string[],
// ): Promise<ElementsDefinition> {
//   if (filter && filter.length > 0) error(501, "bad request");
//   const dg = new DependencyGraph();
//   for (const ws of gprojs) {
//     dg.digestWorkspace(ws);
//   }
//   return dg.elements();
// }

export async function getElements(
  keywordMap: Map<string, PackageJson[]>,
): Promise<ElementsDefinition> {
  const mns = new Map<string, NodeDefinition>();
  const mes = new Map<string, EdgeDefinition>();

  // const nodes = [...packageMap.keys()].map<NodeDefinition>((keyword) => {
  //   return { data: { id: keyword } };
  // });

  keywordMap.forEach((pJs, keyword) => {
    const kwdData: KeywordDataDefinition = getKeywordNodeData(keyword);
    mns.set(kwdData.id, {
      data: kwdData,
      classes: ["keyword"],
    });
    for (const pJ of pJs) {
      const pJData: PackageJsonDataDefinition = getPackageNodeData(pJ);
      mns.set(pJData.id, { data: pJData, classes: ["package"] });

      const edgeData: PkEdgeDefinition = getEdge(kwdData, pJData);
      mes.set(edgeData.id, {
        data: {
          source: kwdData.id,
          target: pJData.id,
        },
      });
    }
  });

  return {
    nodes: Array.from(mns.values()),
    edges: Array.from(mes.values()),
  };

  function getEdge(
    kwdData: KeywordDataDefinition,
    pJData: PackageJsonDataDefinition,
  ): PkEdgeDefinition {
    return {
      id: kwdData.id.concat("|").concat(pJData.id),
      source: kwdData.id,
      target: pJData.id,
    };
  }

  function getPackageNodeData(pJ: PackageJson): PackageJsonDataDefinition {
    const { name, version, description, keywords } = pJ;

    return {
      id: pJ.name.concat("@".concat(pJ.version)),
      name,
      version,
      description,
      keywords,
    };
  }

  function getKeywordNodeData(keyword: string): KeywordDataDefinition {
    return { id: keyword };
  }
}

export function cssDeclarations(): CssStyleDeclaration[] {
  return [
    {
      selector: "node",
      style: { "background-color": "#666" },
    },
    {
      selector: "node.package",
      style: { "background-color": "#f00", label: "data(name)" },
    },
    {
      selector: "node.keyword",
      style: { "background-color": "#0f0" },
    },
    {
      selector: "node.focused",
      style: {
        "background-color": "blue",
        "border-width": 60,
        "border-color": "green",
      },
    },
    {
      selector: "edge",
      style: {
        width: 3,
        "line-color": "#ccc",
        "target-arrow-color": "#ccc",
        "target-arrow-shape": "triangle",
        "curve-style": "bezier",
      },
    },
    {
      selector: "edge.dep.development",
      style: {
        "line-color": "green",
        "target-arrow-color": "green",
      },
    },
    {
      selector: "edge.dep.run-time",
      style: {
        "line-color": "orange",
      },
    },
    {
      selector: ":selected",
      style: {
        "background-color": "yellow",
        "line-color": "yellow",
        "target-arrow-color": "black",
        "source-arrow-color": "black",
      },
    },
    { selector: "edge:selected", style: { width: 20 } },
  ];
}
