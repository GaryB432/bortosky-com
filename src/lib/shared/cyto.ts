import type {
  EdgeDataDefinition,
  EdgeDefinition,
  ElementDataDefinition,
  ElementDefinition,
  ElementsDefinition,
  NodeDataDefinition,
  NodeDefinition,
} from "cytoscape";

interface WorkspaceFeatures {
  angular: boolean;
  devOps: {
    azure?: boolean;
    github?: boolean;
  };
  devkit: boolean;
  management: {
    lerna?: boolean;
    nx?: boolean;
    rush?: boolean;
  };
  rxjs: boolean;
  webpack: boolean;
}

interface NodePackage {
  dependencies?: Record<string, string>;
  description?: string;
  devDependencies?: Record<string, string>;
  homepage?: string;
  keywords?: string[];
  license?: string;
  name: string;
  repository?: unknown;
  version: string;
}

export interface PackageConfiguration {
  features: WorkspaceFeatures;
  json: NodePackage;
  monorepo: unknown[];
  path: string;
}

// export function onep(p: PackageConfiguration): NodeDataDefinition {
//   const pj = p.json;
//   const j = Object.keys(p.json.dependencies ?? {}).map((dep) => ({
//     name: pj.name,
//     dep,
//   }));

//   const qq = j.map((ji) => {
//     return ji;
//   });

//   console.log(j);
//   return {};
// }

function compareElements(a: ElementDefinition, b: ElementDefinition): number {
  const sa = a.data.id ?? "";
  const sb = b.data.id ?? "";
  return sa.localeCompare(sb);
}

export function toElementsDefinition(
  ps: PackageConfiguration[],
): ElementsDefinition {
  const nmap = new Map<string, NodeDefinition>();
  const emap = new Map<string, EdgeDefinition>();

  for (const p of ps) {
    const setEdge = (depk: string, dev: boolean) => {
      if (!nmap.has(depk)) {
        nmap.set(depk, { data: { id: depk } });
      }
      const ek = `${depk}_${p.json.name}`;
      emap.set(ek, {
        data: { id: ek, source: p.json.name, target: depk, dev },
      });
    };

    nmap.set(p.json.name, {
      data: { id: p.json.name, root: true },
      classes: ["root"],
    });
    for (const depk of Object.keys(p.json.devDependencies ?? {})) {
      setEdge(depk, true);
    }
    for (const depk of Object.keys(p.json.dependencies ?? {})) {
      setEdge(depk, false);
    }
  }

  // console.log({ nmap, emap });

  // const nodes = ps.map<NodeDefinition>((p) => {
  //   const { path, features } = p;
  //   // const id = idn.toString();
  //   const id = p.json.name;
  //   const data: NodeDataDefinition = {
  //     id,
  //     path,
  //     features,
  //   };
  //   return { data };
  // });

  // const edges: EdgeDefinition[] = [];
  const nodes = Array.from(nmap.values()).sort(compareElements);
  const edges = Array.from(emap.values()).sort(compareElements);

  edges.forEach(
    (e) => (e.classes = [e.data["dev"] ? "devDependency" : "dependency"]),
  );

  // const f: ElementsDefinition = {
  //   nodes,
  //   edges,
  // };

  // const a = { nodes, edges };

  // console.log(JSON.stringify(a));

  // return JSON.parse(JSON.stringify(a));

  // console.log(JSON.stringify({ nodes, edges }));
  return { nodes, edges };
}
