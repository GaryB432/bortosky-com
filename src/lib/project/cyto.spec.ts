import { describe, expect, test } from "vitest";
import { getDependencyElements, getElements } from "./cyto";
import type { GaryProject } from "./project";

describe("Cyto", () => {
  test("getElements", async () => {
    expect(
      await getElements([
        {
          root: {
            name: "workspace-a",
            version: "0",
            devDependencies: { jest: "3" },
            dependencies: { album: "4" },
          },
          projects: [
            {
              name: "wsa-a",
              type: "module",
              devDependencies: {
                desk: "*",
                window: "*",
                "eslint-plugin-gb": "no-longer-skipped",
              },
              dependencies: { throttle: "4", brake: "1" },
            },
          ],
        },
        {
          root: {
            name: "workspace-b",
            version: "0",
            devDependencies: { jest: "5" },
            dependencies: { album: "2" },
          },
          projects: [],
        },
      ]),
    ).toMatchSnapshot();
  });

  test("getDepencencyElements", async () => {
    expect(
      await getDependencyElements([
        {
          root: {
            name: "workspace-a",
            version: "whatever",
            devDependencies: { jest: "3.0.0" },
            dependencies: { album: "4.0.0" },
          },
          projects: [
            {
              name: "wsa-a",
              type: "module",
              devDependencies: {
                desk: "^1.4.9",
                window: "5.4.2",
                "eslint-plugin-gb": "skipped",
              },
              dependencies: { throttle: "^4.1.2", brake: "~1.4.6" },
            },
          ],
        },
        {
          root: {
            name: "workspace-b",
            version: "0",
            devDependencies: { jest: "^4.0.0" },
            dependencies: { album: "2" },
          },
          projects: [],
        },
      ], ["workspace-a", "workspace-b"]),
    ).toMatchSnapshot();
  });
});

const specimenLarge: GaryProject[] = [
  {
    root: {
      name: "bortosky-com",
      version: "0.0.1",
      type: "module",
      dependencies: {
        tbd: "0.0.0",
      },
      devDependencies: {
        soon: "^17.10.5",
      },
      nx: {},
    },
    projects: [],
  },
  {
    root: {
      name: "district-map",
      version: "0.0.0",
      dependencies: {
        tbd: "0.0.0",
      },
      devDependencies: {
        soon: "^17.10.5",
      },
      workspaces: ["apps/web"],
    },
    projects: [
      {
        name: "reader",
        projectType: "application",
        tags: [],
      },
      {
        name: "reader-e2e",
        implicitDependencies: ["reader"],
      },
      {
        name: "shared",
        projectType: "library",
        tags: [],
      },
      {
        name: "svg",
        projectType: "library",
        tags: [],
      },
      {
        name: "web",
        version: "0.0.1",
        type: "module",
        dependencies: {
          dep: "0",
        },
        devDependencies: {
          devdep: "0",
        },
        nx: {
          implicitDependencies: ["svg", "shared"],
        },
      },
    ],
  },
];
