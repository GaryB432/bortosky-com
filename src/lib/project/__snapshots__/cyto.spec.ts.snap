// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cyto > getElements 1`] = `
{
  "edges": [
    {
      "classes": [
        "dep",
        "run-time",
      ],
      "data": {
        "id": "album_workspace-a",
        "source": "workspace-a",
        "target": "album",
      },
    },
    {
      "classes": [
        "dep",
        "run-time",
      ],
      "data": {
        "id": "album_workspace-b",
        "source": "workspace-b",
        "target": "album",
      },
    },
    {
      "classes": [
        "dep",
        "run-time",
      ],
      "data": {
        "id": "brake_workspace-a#wsa-a",
        "source": "workspace-a#wsa-a",
        "target": "brake",
      },
    },
    {
      "classes": [
        "dep",
        "development",
      ],
      "data": {
        "id": "desk_workspace-a#wsa-a",
        "source": "workspace-a#wsa-a",
        "target": "desk",
      },
    },
    {
      "classes": [
        "dep",
        "development",
      ],
      "data": {
        "id": "jest_workspace-a",
        "source": "workspace-a",
        "target": "jest",
      },
    },
    {
      "classes": [
        "dep",
        "development",
      ],
      "data": {
        "id": "jest_workspace-b",
        "source": "workspace-b",
        "target": "jest",
      },
    },
    {
      "classes": [
        "dep",
        "run-time",
      ],
      "data": {
        "id": "throttle_workspace-a#wsa-a",
        "source": "workspace-a#wsa-a",
        "target": "throttle",
      },
    },
    {
      "classes": [
        "dep",
        "development",
      ],
      "data": {
        "id": "window_workspace-a#wsa-a",
        "source": "workspace-a#wsa-a",
        "target": "window",
      },
    },
    {
      "classes": "psub",
      "data": {
        "id": "workspace-a#wsa-a-workspace-a",
        "source": "workspace-a",
        "target": "workspace-a#wsa-a",
      },
    },
  ],
  "nodes": [
    {
      "classes": [
        "dep",
      ],
      "data": {
        "aversion": "2",
        "id": "album",
      },
    },
    {
      "classes": [
        "dep",
      ],
      "data": {
        "aversion": "1",
        "id": "brake",
      },
    },
    {
      "classes": [
        "dep",
      ],
      "data": {
        "aversion": "*",
        "id": "desk",
      },
    },
    {
      "classes": [
        "dep",
      ],
      "data": {
        "aversion": "5",
        "id": "jest",
      },
    },
    {
      "classes": [
        "dep",
      ],
      "data": {
        "aversion": "4",
        "id": "throttle",
      },
    },
    {
      "classes": [
        "dep",
      ],
      "data": {
        "aversion": "*",
        "id": "window",
      },
    },
    {
      "classes": [
        "gbp",
      ],
      "data": {
        "id": "workspace-a",
      },
    },
    {
      "classes": [
        "psub",
      ],
      "data": {
        "id": "workspace-a#wsa-a",
      },
    },
    {
      "classes": [
        "gbp",
      ],
      "data": {
        "id": "workspace-b",
      },
    },
  ],
}
`;
